### PR TITLE
High level Modify interaction events

### PR DESCRIPTION
**Pull request by sweco-sebhar**
Related to the desired feature as discussed in #2946.

I think this PR needs test cases if the event is to be marked as stable. Also, events are currently triggered without checking for actual changes in the features(s) which probably should be avoided.

### CODE REVIEW

**Comment by sweco-sebhar** [comment]
Check for change introduced in 8ef1550d7a6c2de0ec35bbd7b83b8a74be9fdf6c but it might be useless, there is no case where the geometry would not be modified. In the current master clicks on line segments will raise the geometry revision but no vertex will be created but this is undesirable and should be changed so that vertexes are created on single clicks.

The modify interaction does not currently have any meaningful testcases so this PR will remain untested. Initial testcases should be created as a separate PR.

**Comment by bjornharrtell** [comment]
To clarify the state of this PR I think 8ef1550 should be removed and the behaviour change as described above. I've made tests for the current behavour with #3630 but unfortuneatly it's not trivial to change the behaviour but I intend to try in the near future.

**Comment by bjornharrtell** [comment]
Rebased and squashed without 8ef1550 since I've been able to change the mentioned behaviour with #3635.

**Comment by itjope** [comment]
Would be nice if this gets merged.

**Comment by probins** [comment]
@bjornharrtell could you add `@fires` as in #3740 so it appears in api docs

**Comment by sweco-sebhar** [comment]
Added `@fires`, rebased and squashed.

**Comment by probins** [comment]
I've been thinking about this specifically for implementing a crude split for linestrings. As mentioned in #1165 you could use turfjs to split at an existing vertex. If the vertex at which you want to split doesn't exist, you could add it with the modify interaction, and could then split the line at this new point. But, even if this PR is merged so an app is notified that a change has taken place, the app has no way of knowing which is the new vertex, other than by storing the original list of coordinates and comparing them with the new list - yuk. So I'm thinking that this PR should be changed like interaction.Select to include the mapBrowserEvent, so you know at which pixel/coordinate the new vertex was added.

**Comment by probins** [comment]
alternatively, of course, modify could pass the modified vertex along with a add/change/delete type code

**Comment by sweco-sebhar** [comment]
I think it makes sense to add `mapBrowserEvent` in the same manner as was done for `ol.interaction.Select` in #3757.

**Comment by sweco-sebhar** [comment]
So, I have added `mapBrowserPointerEvent` as a property to ModifyEvent. Noted that `ol.interaction.Modify` uses the more specific MapBrowserPointerEvent instead of MapBrowserEvent that is used for `ol.interaction.Select` and `ol.interaction.Draw`. Don't know yet if there is a good reason for this or not. @probins, please review if possible.

**Comment by bartvde** [comment]
Anybody up for reviewing this today? So it could go in the 3.7 release?

**Comment by probins** [comment]
I think it's fair enough that Modify uses the Pointer event: AIUI, the Pointer event separates mouse down and up (and drag), which you need with Modify, but don't need with Draw; with Draw, you just need the drawn feature, and if you need the start/end coordinate, you can get those from the Geometry.

However, I think the Modify event here needs to include the feature that's being modified. I don't need this myself, as I use it together with Select, and only have 1 feature at a time. But in principle the Modify interaction can have any number of features in the collection, and in this case the app has no way of knowing which feature is being operated on.

That could however be in a separate PR.

File: src/ol/interaction/modifyinteraction.js [hunk 1]
```diff
@@ -27,6 +28,60 @@ goog.require('ol.style.Style');
 
 
 /**
+ * @enum {string}
+ */
+ol.ModifyEventType = {
+  /**
+   * Triggered upon feature modification start
+   * @event ol.ModifyEvent#modifystart
+   * @api stable
```
Comment (bartvde) [inline src/ol/interaction/modifyinteraction.js]: I wouldn't start this as stable? We normally add things first as experimental, and then make them stable later on.

File: src/ol/interaction/modifyinteraction.js [hunk 2]
```diff
@@ -27,6 +28,60 @@ goog.require('ol.style.Style');
...
+   */
+  MODIFYSTART: 'modifystart',
+  /**
+   * Triggered upon feature modification end
+   * @event ol.ModifyEvent#modifyend
+   * @api stable
```
Comment (bartvde) [inline src/ol/interaction/modifyinteraction.js]: same here

File: src/ol/interaction/modifyinteraction.js [hunk 3]
```diff
@@ -27,6 +28,60 @@ goog.require('ol.style.Style');
...
+ * @param {ol.Collection.<ol.Feature>} features The features modified.
+ * @param {ol.MapBrowserPointerEvent} mapBrowserPointerEvent Associated
+ *     {@link ol.MapBrowserPointerEvent}.
+ */
+ol.ModifyEvent = function(type, features, mapBrowserPointerEvent) {
+
+  goog.base(this, type);
+
+  /**
+   * The feature being modified.
```
Comment (bartvde) [inline src/ol/interaction/modifyinteraction.js]: feature or features?

File: src/ol/interaction/modifyinteraction.js [hunk 4]
```diff
@@ -27,6 +28,60 @@ goog.require('ol.style.Style');
...
+   * @type {ol.Collection.<ol.Feature>}
+   * @api stable
```
Comment (bartvde) [inline src/ol/interaction/modifyinteraction.js]: wouldn't start with stable

**Comment by bartvde** [comment]
I put some review comments in just now

**Comment by bjornharrtell** [comment]
Thanks for the review @bartvde and @probins. I agree that this should be experimental api and have fixed the plural/singular typo.

**Comment by bartvde** [comment]
thanks @bjornharrtell I'll merge once Travis gives the okay

**Comment by probins** [comment]
> I think the Modify event here needs to include the feature that's being modified

though there is another way of handling this, namely by adding listeners for change:geometry to the features.

**Comment by probins** [comment]
I've just been testing this in the real world as opposed to simple test data, and find that there is an issue with the pixel/coordinate from the browser event. As said before, you need the coords of the newly created/modified vertex, but because of snapping/tolerance, this probably won't be the same as the coordinate the user clicked: if adding a vertex on a linestring, for example, it's unlikely that the user will click exactly on the line, so any new point will probably not be created at the browser event coord/pixel.

So I'm not sure the browser event is much use here. What the modify event should return is the vertexfeature that's been created/moved/deleted. Should I create a separate issue for this?

**Comment by probins** [comment]
and another issue here is the timing of the events. At the moment, start is before drag but after creating the new vertex. If you are, though, adding a vertex with a click, which you can do now with #3635, there is no drag, so there's effectively little difference between start and end. ISTM in this case you want a start notification at the beginning of the handleDown function, so you get the feature/geometry 'before' modification and then the end event 'after' modification.

When deleting a vertex on the other hand, it looks like the end event is fired before the geometry has been changed.

**Comment by probins** [comment]
> any new point will probably not be created at the browser event coord/pixel

as an aside to this, I thought I might be able to use `getClosestPoint(browserEvent.coordinate)` to get the new vertex, but found that this was a similar but not identical value, so `coordinates.indexOf(...)` didn't find a match. Shouldn't this be the same? Shouldn't creating a new vertex use `getClosestPoint()`?

**Pull request merged**